allow heading content within a legend element

scottaohara opened the pull request:
This PR updates the `legend`’s content model to allow for heading content. This update uses the [`summary` element’s content model](https://html.spec.whatwg.org/#the-summary-element) as the reference for how I think this should be indicated in the spec.

If accepted, this closes #4990

Thanks


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/5090/form-elements.html" title="Last updated on Feb 18, 2020, 10:54 PM UTC (63d4399)">/form-elements.html</a>  ( <a href="https://whatpr.org/html/5090/bcd5d61...63d4399/form-elements.html" title="Last updated on Feb 18, 2020, 10:54 PM UTC (63d4399)">diff</a> )
<a href="https://whatpr.org/html/5090/indices.html" title="Last updated on Feb 18, 2020, 10:54 PM UTC (63d4399)">/indices.html</a>  ( <a href="https://whatpr.org/html/5090/bcd5d61...63d4399/indices.html" title="Last updated on Feb 18, 2020, 10:54 PM UTC (63d4399)">diff</a> )
<a href="https://whatpr.org/html/5090/interactive-elements.html" title="Last updated on Feb 18, 2020, 10:54 PM UTC (63d4399)">/interactive-elements.html</a>  ( <a href="https://whatpr.org/html/5090/bcd5d61...63d4399/interactive-elements.html" title="Last updated on Feb 18, 2020, 10:54 PM UTC (63d4399)">diff</a> )
<a href="https://whatpr.org/html/5090/sections.html" title="Last updated on Feb 18, 2020, 10:54 PM UTC (63d4399)">/sections.html</a>  ( <a href="https://whatpr.org/html/5090/bcd5d61...63d4399/sections.html" title="Last updated on Feb 18, 2020, 10:54 PM UTC (63d4399)">diff</a> )

annevk commented:
https://www.accessibility-developer-guide.com/examples/forms/grouping-with-headings/ suggests this has quite special semantics in its conclusion. In that the level is not announced. Does that need to be reflected somehow?

And when should you use a normal `legend` vs a heading legend?

scottaohara commented:
>suggests this has quite special semantics in its conclusion

the way the conclusion is written is not entirely accurate / robust in exactly how this works.  Because yes, the heading's semantics are not announced when a screen reader navigates into a grouping via form elements quick key or via tab key.... but that's not unique to headings. Any element within a legend will have it's role dropped when the accessible name of the group is announced.

if navigating by heading elements, there is no issue save for a support gap with Firefox and JAWS.  VO + Safari, NVDA + FF/Chrome and JAWS + Chrome all allow for navigating to the headings and will announce the heading role as expected.

So per your question:
>Does that need to be reflected somehow?

no.  it is reflected correctly depending on navigation method, and works as expected save for the one mentioned gap i found in quickly re-testing.

>And when should you use a normal legend vs a heading legend?
when a section of a form should be exposed in the document outline to allow easier access to screen reader users / when a document itself is largely form related and having a separate heading and legend would create duplicative information.  

@stevefaulkner's link to the original discussion w3c/html#724 had a good initial use case by @DavidMacDonald, agreement at the time by @sideshowbarker, and the thread contains various links to issues from gov.uk where they are using this pattern in practice, and are [opting to ignore the current error from the validator](https://github.com/alphagov/govuk-frontend/issues/1080#issuecomment-441996725) due to this working well for them and their users.

annevk commented:
@scottaohara looking at https://github.com/alphagov/govuk_elements/pull/507/files suggests that only allowing a heading element is not enough to make those templates conforming, unless there's some non-obvious post-processing. E.g., some of the `legend` elements there have a heading element followed by a `span` element.

scottaohara commented:
@annevk ah yes. that's an oversight on my part, thank you.  

The intent was to allow a heading as a child of a legend, but to indicate it wouldn't be ideal to have _multiple_ headings.

With that in mind, would one of the following work for you?

Noting they're both allowed, but not being restrictive:
>Phrasing content and heading content.

or rephrasing my original text:

>Either: phrasing content.
>Or: phrasing content and one element of <span>heading content</span>.

Thank you

annevk commented:
I'd expect something like "Phrasing content, optionally intermixed with heading content" (not sure restricting to a single element makes sense if you allow `hgroup`). I guess the other question is whether the `summary` element should similarly change.

scottaohara commented:
I've adjusted the PR to match the wording you provided, thank you.

Regarding updating `summary`, would you want that done in this PR, or discuss in a separate issue?

sideshowbarker commented:
Note that the original discussion at https://github.com/w3c/html/issues/724 was about allowing each `legend` element to contain either heading content or phrasing content, but not both — not intermixed.

> looking at https://github.com/alphagov/govuk_elements/pull/507/files suggests that only allowing a heading element is not enough to make those templates conforming, unless there's some non-obvious post-processing. E.g., some of the `legend` elements there have a heading element followed by a `span` element.

If @scottaohara and others who’ve looked in detail at that site and related cases are confident there’s an accessibility win in allowing a `legend` element to contain an intermixing of heading-content children and phrasing-content children — rather than an either-or that allows each `legend` to only have one or the other — then yeah, "Phrasing content, optionally intermixed with heading content" is the right wording. Otherwise, I’d suggest we restrict it to being either-or.

> I guess the other question is whether the `summary` element should similarly change.

Yeah, since we’re allowing `hgroup` in summary, then I think the `summary` content model should be changed to match what’s in the patch here for `legend` ("Phrasing content, optionally intermixed with heading content").

> Regarding updating `summary`, would you want that done in this PR, or discuss in a separate issue?

IMHO, we should include the `summary` change here. That way, somebody reading through the change history later will see that these were changed together, with the same rationale.

scottaohara commented:
Thank you both for your feedback here.  I've updated the PR now to adjust the content model for `summary` to match the updated model for `legend`.

annevk commented:
I was going to land this as `Allow heading content within legend and summary elements` but I realized something:

1. We need to update h1-6 and hgroup to state that they can be used "Where heading content is expected."
2. Maybe we should add an example?

36degrees commented:
Hi folks,

Exciting to see this change going through.

I just wanted to flag that in a recent audit of the GOV.UK Design System some bugs with JAWS and ZoomText were reported when the heading was placed inside a legend.

https://github.com/alphagov/govuk-frontend/issues/1605

As discussed in the issue, this appears to affect JAWS 2018 and above – JAWS 17/18 which we previously tested with when we made the change work fine. We've reported those issues to Freedom Scientific. We haven't yet been able to reproduce the issue reported in ZoomText.

We still believe that allowing headings inside legends is the right approach. Hopefully having this in the WHATWG HTML spec will help to get things fixed when we're reporting bugs with the vendors.

scottaohara commented:
hi @36degrees 
yes i noticed the JAWS issue as well and mentioned in the related issue to this PR. I've passed along the information and a bug with JAWS will be filed.

sideshowbarker commented:
@scottaohara I think this PR is waiting on the further changes needed that @annevk listed in https://github.com/whatwg/html/pull/5090#issuecomment-558751618

scottaohara commented:
@sideshowbarker @annevk, apologies.  it's been on my to do list.  I'll try to get it done within the next few days.

scottaohara commented:
@annevk i believe i've addressed your bullets.  i added an example to the end of the other fieldset examples.  if you think that'd be best somewhere else, or if you notice anything i've missed/would like additional updates on, just let me know.

thanks

annevk commented on source:
```diff
@@ -119595,6 +119619,7 @@ interface <dfn>External</dfn> {
          <span data-x="Heading content">heading</span>;
          <span data-x="Palpable content">palpable</span></td>
      <td><code>hgroup</code>;
+         <code>legend</code>;
```
Should summary be listed here? Arguably hgroup should not be since it's part of flow.

annevk replied:
Let's flip the order of summary/legend here so it's consistent with the later one? Nothing else from me. Hopefully @sideshowbarker can look tomorrow.

scottaohara replied:
done

annevk commented on source:
```diff
@@ -119628,7 +119653,9 @@ interface <dfn>External</dfn> {
          <span data-x="Palpable content">palpable</span></td>
-     <td><span data-x="Flow content">flow</span></td>
+     <td>
+      <code>legend</code>;
```
Should summary be listed here?

scottaohara commented:
thanks @annevk.  updated per your comments and added heading content to the summary index.

domenic commented on source:
```diff
@@ -52670,6 +52670,31 @@ interface <dfn>HTMLFieldSetElement</dfn> : <span>HTMLElement</span> {
+   <pre><code class="html" data-x="">&lt;fieldset>
+ &lt;legend> &lt;h1>
+  How can we best reach you?
+ &lt;/h1> &lt;/legend>
```
Given that multiple h1s per document is hopefully going away (https://github.com/whatwg/html/pull/3499#issuecomment-577866180), perhaps we should use something other than h1 here?

scottaohara replied:
updated.  thanks

annevk commented:
Thanks @scottaohara!
